### DISCUSSION TITLE
Why is the Java runtime partially thread safe?

### DISCUSSION THREAD

**Discussion by lppedd** [antlr/antlr4#4508] [Q&A]
Rewriting the Kotlin runtime taking the Java runtime as starting point, I've noticed that here and there it uses thread-safe utilities like synchronized blocks or `CopyOnWriteArrayList`s.

Looking for answers to "Is ANTLR thread safe?", we [find](https://stackoverflow.com/a/19329167/1392277):

> In ANTLR 4, instance methods are not safe for multi-threaded operations, but you may create two separate instances of the parser and use those separate instances on multiple threads

So what's the point of using, for example, a `CopyOnWriteArrayList`?  
What's the ANTLR team opinion on this subject?

**Comment by KvanTTT** [comment]
It could be just a bug or overengineering.

Some time ago I enountered that sometimes ANTLR is freezing in multithread mode. I started investigating and noticed that IDEA [warns](https://github.com/antlr/antlr4/issues/3756) that some code may cause deadlock. I [fixed](https://github.com/antlr/antlr4/commit/875911c2b2e6e50317e75ff2b5e80f4014a75954) the code and it fixed the original problem.
